Add a simple formatting helper

randombit opened the pull request:
This can serve as a standin for `std::format` until either a more `std::format` like helper can be devised or (longer term) we can switch to a compiler provided `std::format`.

randombit commented:
I just realized it's very easy to make this `std::format` compatible using `{}` but sadly I can't push it because GH is having problems.

reneme commented:
FWIW: Also had trouble with GitHub half an hour ago. Though, at least from Berlin, it works again.

reneme commented on src/lib/utils/fmt.h:
```diff
@@ -0,0 +1,68 @@
+   {
+   size_t i = 0;
+
+   while(i < format.size())
+      {
+      if(format[i] == '{' && format[i + 1] == '}')
+         {
+         oss << val;
+         return do_fmt(oss, format.substr(i + 2), rest...);
+         }
+      else
+         {
+         oss << format[i];
+         }
+
+      i += 1;
+      }
+   }
```
There's an off-by-one here, no?

Consider `format = "I will make the sanitizer unhappy {"`. That will cause an out-of bounds read when looking for the closing brace.

```suggestion
   while(i < format.size() - 1)
      {
      if(format[i] == '{' && format[i + 1] == '}')
         {
         oss << val;
         return do_fmt(oss, format.substr(i + 2), rest...);
         }
      else
         {
         oss << format[i];
         }

      i += 1;
      }
   }
```

randombit replied:
Fixed now, also added tests

reneme commented on src/lib/pubkey/pbes2/pbes2.cpp:
```diff
@@ -85,15 +87,15 @@ secure_vector<uint8_t> derive_key(std::string_view passphrase,
-      throw Decoding_Error("PBE-PKCS5 v2.0: Unknown KDF algorithm " +
-                           kdf_algo.oid().to_string());
+      {
+      throw Decoding_Error(fmt("PBE-PKCS5 v2.0: Unknown KDF algorithm {}",
+                                kdf_algo.oid().to_string()));
```
As a side note: When we eventually move to `std::format` these `to_string()` gymnastics should be handled centrally with our custom types IMO. Didn't look into it in detail, but I'd be surprised if one could not define this generically for user-defined types in `std::format`.

For now, I think, this is totally sufficient.

randombit replied:
I haven't looked into in detail either (seems somewhat academic without a compiler to test anything with) but it appears [std::formatter](https://en.cppreference.com/w/cpp/utility/format/formatter) will let us add custom flags support and other fancy stuff.

For the short term I added an operator<< for OID, which is sufficient to avoid the explicit `to_string` calls, since our `fmt` will happily eat anything it can put into an `ostream`

reneme reviewed (approved):
Nice! LGTM, modulo the potential out-of-bounds read and maybe a few basic unit tests.

randombit commented:
Fixed the off by one, added some unit tests and also expanded usage somewhat to other places where we were doing string concatenations, especially `"Foo(" + x + ")"`. Adopting a parenthesized notation for algorithm names should probably go into [`mistakes.rst`](https://github.com/randombit/botan/blob/master/doc/dev_ref/mistakes.rst) 

I was a little worried adopting this widely was going to blow up the binary size but happily it seems like (at least on my machine/compiler) the resulting binary actually got a bit smaller, vs `master`.
